Add e2e tests for publishing plug-ins

ScalaWilliam opened the pull request:
Add 2 scripted-level tests. Expectation in the tests follows:
- If we publish a plug-in, it should be accessible in a normal standard way
- If we publish a plug-in, it should be also accessible in a consistent way

This assumes that the publishing is done in a __single__ step in both cases (see below for the reason why).

Running the test `scripted write/consistent` showed:

```
[info] [info] Created a proper POM file: C:\Users\XYZ\AppData\Local\Temp\sbt_959e3e42\testPlugin\target\scala-2.12\sbt-1.0\test-plugin_sbt1_2.12.pom
[info] [info]   published test-plugin_sbt1_2.12 to file:/C:/Users/XYZ/.m2/repository/com/test/test-plugin_2.12_1.0/0.0.2-SNAPSHOT/test-plugin_sbt1_2.12-0.0.2-SNAPSHOT.jar
[info] [info]   published test-plugin_sbt1_2.12 to file:/C:/Users/XYZ/.m2/repository/com/test/test-plugin_2.12_1.0/0.0.2-SNAPSHOT/test-plugin_sbt1_2.12-0.0.2-SNAPSHOT-javadoc.jar
[info] [info]   published test-plugin_sbt1_2.12 to file:/C:/Users/XYZ/.m2/repository/com/test/test-plugin_2.12_1.0/0.0.2-SNAPSHOT/test-plugin_sbt1_2.12-0.0.2-SNAPSHOT-sources.jar
[info] [info]   published test-plugin_sbt1_2.12 to file:/C:/Users/XYZ/.m2/repository/com/test/test-plugin_2.12_1.0/0.0.2-SNAPSHOT/test-plugin_sbt1_2.12-0.0.2-SNAPSHOT.pom
Expected -->                                             C:\Users\XYZ\.m2\repository\com\test\test-plugin_sbt1_2.12\0.0.2-SNAPSHOT\test-plugin_sbt1_2.12-0.0.2-SNAPSHOT.pom

([info] [error]   not found: https://repo1.maven.org/maven2/com/test/test-plugin_sbt1_2.12/0.0.2-SNAPSHOT/test-plugin_sbt1_2.12-0.0.2-SNAPSHOT.pom)
```

Noticing here a couple of things: only the filename is affected, but it does not match the artifact ID (small change to fix this); the standard version is not published either.

This raises 2 important questions:

1. Do we publish consistent & inconsistent plug-ins in one command or two commands?
2. Should the artifactId be the same for both (ie. do we get two directories of 4 files each, or 1 directory of 8 published files)?

For sbt-assembly, Eugene Yokota published a convention of `sbt1_2.12` and published the module in a completely separate command.

My personal preference would be to:
- Publish in 1 command, so that plug-in authors don't have to think too much, especially from their CI pipelines.
- Publish with the same artifactId, so it is obvious that there are 2 different types of poms in the distribution; also this should simplify publishing to sonatype as you'd be publishing 1 artifactId.

Therefore, we'd need to change to _append_ rather than rewrite the list of artifacts; and also remove the sbt1_2.12 convention.

ScalaWilliam committed:
Add e2e tests for publishing the plug-in.

Running the test `scripted write/consistent` showed:

```
[info] [info] Created a proper POM file: C:\Users\XYZ\AppData\Local\Temp\sbt_959e3e42\testPlugin\target\scala-2.12\sbt-1.0\test-plugin_sbt1_2.12.pom
[info] [info]   published test-plugin_sbt1_2.12 to file:/C:/Users/XYZ/.m2/repository/com/test/test-plugin_2.12_1.0/0.0.2-SNAPSHOT/test-plugin_sbt1_2.12-0.0.2-SNAPSHOT.jar
[info] [info]   published test-plugin_sbt1_2.12 to file:/C:/Users/XYZ/.m2/repository/com/test/test-plugin_2.12_1.0/0.0.2-SNAPSHOT/test-plugin_sbt1_2.12-0.0.2-SNAPSHOT-javadoc.jar
[info] [info]   published test-plugin_sbt1_2.12 to file:/C:/Users/XYZ/.m2/repository/com/test/test-plugin_2.12_1.0/0.0.2-SNAPSHOT/test-plugin_sbt1_2.12-0.0.2-SNAPSHOT-sources.jar
[info] [info]   published test-plugin_sbt1_2.12 to file:/C:/Users/XYZ/.m2/repository/com/test/test-plugin_2.12_1.0/0.0.2-SNAPSHOT/test-plugin_sbt1_2.12-0.0.2-SNAPSHOT.pom
Expected -->                                             C:\Users\XYZ\.m2\repository\com\test\test-plugin_sbt1_2.12\0.0.2-SNAPSHOT\test-plugin_sbt1_2.12-0.0.2-SNAPSHOT.pom

([info] [error]   not found: https://repo1.maven.org/maven2/com/test/test-plugin_sbt1_2.12/0.0.2-SNAPSHOT/test-plugin_sbt1_2.12-0.0.2-SNAPSHOT.pom)
```

Noticing here a couple of things: only the filename is affected, but it does not match the artifact ID (small change to fix this); the standard version is not published either.

This raises 2 important questions:

1. Do we publish consistent & inconsistent plug-ins in one command or two commands?
2. Should the artifactId be the same for both (ie. do we get two directories of 4 files each, or 1 directory of 8 published files)?

For sbt-assembly, Eugene Yokota published a convention of `sbt1_2.12` and published the module in a completely separate command.

My personal preference would be to:
- Publish in 1 command, so that plug-in authors don't have to think too much, especially from their CI pipelines.
- Publish with the same artifactId, so it is obvious that there are 2 different types of poms in the distribution; also this should simplify publishing to sonatype as you'd be publishing 1 artifactId.

Therefore, we'd need to change to _append_ rather than rewrite the list of artifacts; and also remove the sbt1_2.12 convention.

ScalaWilliam commented:
@wudong the build should fail now - would you take a crack at making it pass? It would require some adjustments in the publishing aspect (although it is already super close).

If you agree with the thinking please merge this PR.

wudong commented:
Using the `sbt1_2.12` convention seems conflicting with how sbt cross the name, it would cross the name with `scalaBinaryVersion_sbtBinaryVersion`

That's why currently we are seeing a very strange pattern which have a mix of both:

`com/test/test-plugin_2.12_1.0/0.0.2-SNAPSHOT/test-plugin_sbt1_2.12-0.0.2-SNAPSHOT.jar`

with the `test-plugin_2.12_1.0` to be a crossed result done by sbt from the module name, while we have overridden the artifact name.

Personally I think it would be better to be consistent with sbt's own conversion, and make it to be: 
`com/test/test-plugin_2.12_1.0/0.0.2-SNAPSHOT/test-plugin_2.12_1.0-0.0.2-SNAPSHOT.jar`

or it is a bit more of a hacking to be done:
1. disable crossing on the moduleName,
2. override the moduleName (which i'm not sure if it would have a wider impact).

ScalaWilliam commented:
> Personally I think it would be better to be consistent with sbt's own conversion, and make it to be:
> `com/test/test-plugin_2.12_1.0/0.0.2-SNAPSHOT/test-plugin_2.12_1.0-0.0.2-SNAPSHOT.jar`

yep. that.

wudong committed:
Publishing both consistent POM and original POM.

1. change to use the sbt convention for crossname
2. change to not use random version in scripted test, it seems each step in the test would have a new random number generated, which causing the test to fail.

wudong commented:
There is also problem at publishing in both ways at the same time.. 

We actually cannot publish in the `normal` way to artifactory, with the POM being invalid.. 

It would need to be a decision / setting done by the developer depending on which repository they want to publish to. If publishing to artifactory, they can only publish in consistent way.

I think probably best way forward is a setting which allow developer to choose:
1. publish consistently only -> for case of artifactory
2. publish both normal / consistently -> for case that the POM check isn't done.

ScalaWilliam commented:
> There is also problem at publishing in both ways at the same time..
> 
> We actually cannot publish in the `normal` way to artifactory, with the POM being invalid..
> 
> It would need to be a decision / setting done by the developer depending on which repository they want to publish to. If publishing to artifactory, they can only publish in consistent way.
> 
> I think probably best way forward is a setting which allow developer to choose:
> 
> 1. publish consistently only -> for case of artifactory
> 2. publish both normal / consistently -> for case that the POM check isn't done.

In what way is it not possible, what error message do you see?

wudong commented:
I think artifactory doesn't not allow publish  with invalid POM, isn't that the problem we faced within enterprise cause they dont want to relax on this? 
I got a 409 error when trying to publish the 'normal' POM.

ScalaWilliam commented:
> I think artifactory doesn't not allow publish with invalid POM, isn't that the problem we faced within enterprise cause they dont want to relax on this? I got a 409 error when trying to publish the 'normal' POM.

Yes, publishing invalid against Artifactory will generally not work unless an explicit POM_CONSISTENCY option is set to 0, but this will not matter.

However we only need to publish this to Sonatype OSS (Nexus), which accepts invalid POM. When enterprise Artifactory fetches from Sonatype/Maven Central, it will only fetch the valid POM. I just don't know about _publishing_ 2 POM to Nexus, though I expect that it would work because it already accepts _invalid_ in the first place.

wudong commented:
Alright, I have updated it now. please have a look.

ScalaWilliam commented:
> Alright, I have updated it now. please have a look.

@wudong looks good. I got this output:

```
[info] [info] Created a proper POM file:                C:\Users\SampleUser\AppData\Local\Temp\sbt_7fbf04b2\testPlugin\target\scala-2.12\sbt-1.0\test-plugin_2.12_1.0.pom
[info] [info]   published test-plugin_2.12_1.0 to file:/C:/Users/SampleUser/.m2/repository/com/test/test-plugin_2.12_1.0/0.0.312345156/test-plugin_2.12_1.0-0.0.312345156-sources.jar
[info] [info]   published test-plugin to          file:/C:/Users/SampleUser/.m2/repository/com/test/test-plugin_2.12_1.0/0.0.312345156/test-plugin-0.0.312345156-sources.jar
[info] [info]   published test-plugin_2.12_1.0 to file:/C:/Users/SampleUser/.m2/repository/com/test/test-plugin_2.12_1.0/0.0.312345156/test-plugin_2.12_1.0-0.0.312345156-javadoc.jar
[info] [info]   published test-plugin to          file:/C:/Users/SampleUser/.m2/repository/com/test/test-plugin_2.12_1.0/0.0.312345156/test-plugin-0.0.312345156.pom
[info] [info]   published test-plugin_2.12_1.0 to file:/C:/Users/SampleUser/.m2/repository/com/test/test-plugin_2.12_1.0/0.0.312345156/test-plugin_2.12_1.0-0.0.312345156.jar
[info] [info]   published test-plugin_2.12_1.0 to file:/C:/Users/SampleUser/.m2/repository/com/test/test-plugin_2.12_1.0/0.0.312345156/test-plugin_2.12_1.0-0.0.312345156.pom
[info] [info]   published test-plugin to          file:/C:/Users/SampleUser/.m2/repository/com/test/test-plugin_2.12_1.0/0.0.312345156/test-plugin-0.0.312345156-javadoc.jar
[info] [info]   published test-plugin to          file:/C:/Users/SampleUser/.m2/repository/com/test/test-plugin_2.12_1.0/0.0.312345156/test-plugin-0.0.312345156.jar
[info] [success] Total time: 2 s, completed 7 Jul 2022, 19:19:17
[info] [success] Total time: 0 s, completed 7 Jul 2022, 19:19:17
```

Will try to publish a real plug-in and consume it through the consistent approach.
